modules/ROOT/pages/test-mule-applications.adoc

```diff
@@ -0,0 +1,61 @@
+
+See xref:profiling-mule.adoc[Profiling Mule] for more instructions.
+
+== Unit Tests
+
+The concept of unit tests may vary according to the programming paradigm that you are working on, but it always points to the same core concept: _to validate the correctness of an individual unit of source code_.
+
+In the context of unit testing, a unit of code is the smallest testable part of an application. What exactly constitutes the smallest testable part of an application depends on the application. When testing Mule applications, MuleSoft considers that the smallest testable part is a Mule flow or a sub-flow.
+
+Test cases must be independent of each other and it is the developer’s responsibility to ensure test inter-independency.
+The code you test might interact with other units of code or components, always consider this possibility when designing unit tests. A good unit test ensures the isolation of the unit of code being tested, to avoid mistaking failure in other components for failure in the unit of code being tested.
+
+To isolate the target unit of code, use tools such as the `mock-when` processor provided by MUnit.
+
+See xref:munit::mock-before-after-foreach-cookbook.adoc[Mocking a Message Before and Inside a Foreach Processor] for a usage example of the `mock-when` processor.
+
+== Integration Tests
+
+Units of code collaborate between them to create an actual application. You test individual units of code with unit tests, and you test how units of code collaborate between them with integration tests.
+
+The goal of an Integration test is _to validate that different units of code and modules work together as intended_.
+Integration tests must aggregate units of code modules that have already been successfully unit-tested. For this reason, always run your integration tests after your unit tests.
+
+Depending on the nature of your application, integration tests may require sandboxes, that is, environments for your application to connect to. When working with sandboxes, ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.
+
+== Writing Testable Code
+
+The following guidelines help you create flows that are easier to test. Writing testable code implies creating a more configurable, extensible, and readable application.
+
+* Modularize your code
++
+Breaking down your code in different files helps with readability. Do this by grouping flows that work together to achieve a common goal, or group code according to specific functional criteria.
+* Write short flows
++
+Long flows are often hard to follow, harder to code, and harder to maintain. From the perspective of unit testing, long flows offer too many scenarios that can be triggered by a single point. Long flows might require you to perform very complex evaluations to validate a single scenario and are best avoided.
```

Comment:
```suggestion
Long flows can be hard to follow, hard to code, and hard to maintain, so it is best to avoid them. From the perspective of unit testing, long flows offer too many scenarios that can be triggered by a single point, and they require you to perform very complex evaluations to validate a single scenario.
```